### DISCUSSION TITLE
Entities with inheritance causing problems

### DISCUSSION THREAD

**Discussion by vknht** [nhibernate/nhibernate-core#3626] [General]
Hello,

i got an Entity "FooTableExEntity" which inherits from "FooTableEntity".
So far so good.

If i QueryOver of the FooTableEntity-Type for a specific Id, i get 2 results (one for each Type of Entities).

![grafik](https://github.com/user-attachments/assets/a41a1313-0bc6-4d2f-8865-485ee6933af2)

I am not sure if this is a Bug, or a mapping Error.
Attached is a minimal Project to reproduce the behaviour including the Sql-Script.

If QueryOver of the FooTableExEntity-Type i only get 1 result.

NHibernate: 5.5.2
FluentNhibernate: 3.4.0

[EntityInheritanceTest.zip](https://github.com/user-attachments/files/17645202/EntityInheritanceTest.zip)

**Comment by hazzik** [comment]
Hi, you've mapped your classes as two separate entities, not as subclasses. 

So:  `public class FooTableExMap : ClassMap<FooTableExEntity>` should be `public class FooTableExMap : SubclassMap<FooTableExEntity>`

**Comment by vknht** [comment]
Hi,

i tried that but i am getting this Error:
"Cannot map joined-subclass EntityInheritanceTest.FooTableExEntity to table FooTable, the same table as its base class."

If I remove "Table("FooTable")" from the mapping (FooTableExMap) i get the following error:
"Invalid object name 'dbo.FooTableExEntity'."

The criteria is still the same:
```C#
var crit = QueryOver.Of<FooTableEntity>()
                    .Where(x => x.Id == 1)
                    .DetachedCriteria;
```

**Comment by caccorsi** [comment]
It should be like this, if the table is the same in FooTableEx, don't specify. 

    public class FooTableExMap : SubclassMap<FooTableExEntity>
    {
        public FooTableExMap()
        {
            Map(x => x.Name, "Name");
        }
    }

**Comment by vknht** [comment]
Hello,

like i said before, if i remove the Table statement, i get this error:
"Invalid object name 'dbo.FooTableExEntity'."

![grafik](https://github.com/user-attachments/assets/641b0cd9-7539-4cd0-9e42-a5cba2227a8f)

this is the class/mapping:

```csharp
using FluentNHibernate.Mapping;

namespace EntityInheritanceTest
{
    public class FooTableExEntity : FooTableEntity
    {
        public virtual string Name { get; set; }
    }

    public class FooTableExMap : SubclassMap<FooTableExEntity>
    {
        public FooTableExMap()
        {
            //Table("FooTable");
            Map(x => x.Name, "Name");
        }
    }
}

```

**Reply by caccorsi** [reply]
Start with an empty database. You can limit the grants however you wish in the future, but to make things simple, make sure the db_owner role is assigned to the credentials you're using. 

These two maps worked for me.
Note:  'Key' is a reserved work in SQL Server, use something else. 

    public class FooTableMap : ClassMap<FooTableEntity>
    {
        public FooTableMap()
        {
            Table("FooTable");
            DiscriminateSubClassesOnColumn("Type");
            Id(x => x.Id, "Id");
            Map(x => x.Key, "KeyCol");
            Map(x => x.Value, "ValueCol");
        }
    }

and 

    public class FooTableExMap : SubclassMap<FooTableExEntity>
    {
        public FooTableExMap()
        {
            Map(x => x.Name, "Name");
        }
    }

You use a discriminator to differentiate the subtypes. NH takes care of that although you can customize the discriminator values. 
If you want separate tables for each type, that's a whole other discussion. Good luck!

![nh_issue](https://github.com/user-attachments/assets/02b1cb4f-e976-462f-b628-6bd33c17031b)

**Comment by oskarb** [comment]
If you have objects of different types in the same table, how can NHibernate know which type it is? For this, you need to configure discriminators. The core documentation: https://nhibernate.info/doc/nhibernate-reference/mapping.html#mapping-declaration-discriminator  (It's also available in FluentNHibernate).

If you "track" object type by keeping subclasses in different tables (i.e. "joined-subclass") the discriminator is not necessary but in this case it doesn't make sense to try and force the same table to join with itself.

Also, beware of any conventions you might have setup that messes with it.
